pkg/controller/endpointslice/endpointslice_controller.go

```diff
@@ -343,6 +355,49 @@ func (c *Controller) onServiceDelete(obj interface{}) {
 	c.queue.Add(key)
 }
 
+// onEndpointSliceAdd queues a sync for the relevant Service for a sync if the
+// EndpointSlice resource version does not match the expected version in the
+// endpointSliceTracker.
+func (c *Controller) onEndpointSliceAdd(obj interface{}) {
+	endpointSlice := obj.(*discovery.EndpointSlice)
+	if managedByController(endpointSlice) && !c.endpointSliceTracker.UpToDate(endpointSlice) {
+		c.queueServiceForEndpointSlice(endpointSlice)
+	}
+}
+
+// onEndpointSliceUpdate queues a sync for the relevant Service for a sync if the
+// EndpointSlice resource version does not match the expected version in the
+// endpointSliceTracker or the managed-by value of the EndpointSlice has changed
+// from or to this controller.
+func (c *Controller) onEndpointSliceUpdate(prevObj, obj interface{}) {
+	prevEndpointSlice := obj.(*discovery.EndpointSlice)
+	endpointSlice := obj.(*discovery.EndpointSlice)
+	if managedByChanged(prevEndpointSlice, endpointSlice) || (managedByController(endpointSlice) && !c.endpointSliceTracker.UpToDate(endpointSlice)) {
+		c.queueServiceForEndpointSlice(endpointSlice)
```

Comment:
if the old and new endpoint slice were both managed by us, but the service they referenced changed, don't we need to queue the old and new service?

Reply:
That's not a use case I'd planned for. The controller itself will never change the Service an EndpointSlice belongs to. Of course this could happen if someone were intentionally messing with EndpointSlices, but they could also break other things as well. With that said, the `UpToDate` check would fail if the EndpointSlice service had been changed since it would not expect that EndpointSlice name or resource version for the given Service.